Add documentation for trace

turkenh opened the pull request:
## Overview

This PR:

* Adds a separate documentation for trace command.
* Adds a short information about the trace command to main readme as well as a reference to the seperate documentation.

turkenh committed:
doc: add documentation for trace command

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
doc: Update readme for trace command.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

suskin commented on README.md:
````diff
@@ -60,6 +62,30 @@ kubectl crossplane stack list
+It finds the relevant objects for a given one and provides detailed information.
+
+```
+$ kubectl crossplane trace TYPE[.GROUP] NAME [-n| --namespace NAMESPACE]
````
Recently we've moved away from putting the `$` shell prompt at the beginning of the line so that it's easier to copy and paste blindly into a terminal session. Would you mind removing the leading `$` characters from the code examples?

suskin commented on docs/trace-command.md:
````diff
@@ -0,0 +1,117 @@
+
+```bash
+$ kubectl crossplane trace --help
+Usage: kubectl crossplane trace TYPE[.GROUP] NAME [-n| --namespace NAMESPACE] [-h|--help]
````
I usually try to avoid putting the full output of `--help` in the documentation so that it doesn't need to be maintained, but I could go either way in this case

suskin commented on docs/trace-command.md:
```diff
@@ -0,0 +1,117 @@
+    - listing/filtering all `KubernetesApplicationResources` and `ResourceClaims` in the same namespace with labels 
+  at `spec.resourceSelector.matchLabels`
+  
+- Kubernetes Application Resource 
```
Do we have a link for this one?

turkenh replied:
No dedicated section for this in concepts doc. But linked to workloads section which sounds closest.

suskin commented on docs/trace-command.md:
```diff
@@ -0,0 +1,117 @@
+    - `status.clusterRef` -> KubernetesCluster
+    - `spec.secrets` -> [Connection secrets](https://github.com/crossplaneio/crossplane/blob/master/docs/concepts.md#connection-secrets)
+   
+- [Resource Claim](https://github.com/crossplaneio/crossplane/blob/master/docs/concepts.md#kubernetes-application)
```
I believe this link should be more like `https://github.com/crossplaneio/crossplane/blob/master/docs/concepts.md#resource-claim`

suskin commented on docs/trace-command.md:
```diff
@@ -0,0 +1,117 @@
+graph is outputted either in text format or as [graph definition dot format](https://www.graphviz.org/doc/info/lang.html). 
+Text output contains both an overview of all related types indicating [their state](#state-of-a-resource))
+
+## Future Work
```
Nice - I like that we included this section to give people a preview of what we have in mind

suskin commented on docs/trace-command.md:
````diff
@@ -0,0 +1,117 @@
+$ kubectl crossplane trace KubernetesApplication wordpress-app-83f04457-0b1b-4532-9691-f55cf6c0da6e -n app-project1-dev -o dot | dot -Tpng > /tmp/output.png
+```
+
+### State of a resource
````
Some of the content in this document (especially the examples and the explanation of the state characters) seem like they'd be useful to have in the output of `--help`. Would we be able to add that content to the output of `--help`?

turkenh replied:
That is definitely a good idea. Done, Thanks!

suskin commented on README.md:
````diff
@@ -60,6 +62,30 @@ kubectl crossplane stack list
+$ kubectl crossplane trace KubernetesApplication wordpress-app-83f04457-0b1b-4532-9691-f55cf6c0da6e -n app-project1-dev -o dot | dot -Tpng > /tmp/output.png
+```
+
+For more information, see [the trace command documentation](docs/trace-command.md).
````
Good call adding a detailed doc

suskin commented on docs/trace-command.md:
````diff
@@ -0,0 +1,117 @@
+The trace tool is part of the crossplane kubectl plugin and can be used as follows:
+
+```bash
+$ kubectl crossplane trace --help
````
Maybe this is in scope for this since it's a documentation pull request, but @displague reports that the tool displays the following output if no arguments are provided:

```
$ kubectl crossplane trace
missing arguments: SUBCOMMAND TYPE[.GROUP] NAME [-n|--namespace NAMESPACE]
```

Would we be able to use `trace` instead of `SUBCOMMAND`, or just print out the full usage if required arguments are missing?

turkenh replied:
Done

suskin reviewed (commented):
Thanks @turkenh ! Good call on adding a summary to the main README and a longer treatment in a separate doc

suskin commented:
> The main readme file still needs to be updated for new build/install process

@turkenh Now that #17 is merged, what do you think of whether we're able to include that in this changeset too?

turkenh committed:
doc: remove $ signs from code examples

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
doc: fix links to Crossplane resources

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: update help message

Include more details in the help message like examples and the
meanings of signs in state column.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
installation: update bootstrap.sh and readme.md for releases

Installation script and documentation were updated to reflect
recent changes regarding distibuting the tool.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh commented:
@suskin thanks for the review. 
All should be resolved now, including updates for new build/install process.

turkenh committed:
bootstrap: enable release-0.1 source install for backward compat

Enables installation from source for "RELEASE=release-0.1" for
backward compatibility.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

suskin commented on cmd/trace/main.go:
```diff
@@ -105,7 +104,26 @@ func failWithMessage(msg string) {
+| Missing  |          ✕          |     Dotted - Red     |
+ ------------------------------------------------------- 
+
+Usage: kubectl crossplane trace TYPE[.GROUP] NAME [-n| --namespace NAMESPACE] [-h|--help]
```
We should put the usage line at the top so it matches the way manpages order things

turkenh replied:
Well actually, I preferred this way to be inline with kubectl commands, i.e. `kubectl get --help`
What I see is `man` shows top section of the help message first, but when one writes `--help` it is different. 
What do you think?

suskin replied:
Ahh, ok, that makes sense. I didn't realize `kubectl` does it that way for `--help`, and I like the idea of matching `kubectl`. This looks good to me then, so I'll approve it!

suskin commented on bootstrap.sh:
```diff
@@ -7,7 +7,7 @@ if [[ -z "${PREFIX}" ]]; then
 
 if [[ -z "${RELEASE}" ]]; then
-  RELEASE=master
+  RELEASE=$(curl -L -s -H 'Accept: application/json' https://github.com/crossplaneio/crossplane-cli/releases/latest | sed -e 's/.*"tag_name":"\([^"]*\)".*/\1/')
```
This is awesome - great idea! I tested this locally and discovered that `latest` doesn't include releases marked as pre-releases, so we'll want to keep that in mind when we're cutting our releases.

suskin reviewed (changes requested):
Requesting changes for moving the usage line to the first line of the help output. Otherwise, looks good to me!
